File: leetcode/0108-convert-sorted-array-to-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        return self.sortedArrayToBSTUtil(nums, 0, len(nums) - 1)
    
    def sortedArrayToBSTUtil(self, nums: List[int], left: int, right: int) -> TreeNode:
        if right < left:
            return None
        
        if left == right:
            return TreeNode(nums[left])
        
        mid = left + (right - left) // 2
        newNode = TreeNode(nums[mid])
        newNode.left = self.sortedArrayToBSTUtil(nums, left, mid - 1)
        newNode.right = self.sortedArrayToBSTUtil(nums, mid + 1, right)
        
        return newNode

File: leetcode/0108-convert-sorted-array-to-binary-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return sortedArrayToBST(nums, 0, nums.size()-1);
    }
    
    TreeNode* sortedArrayToBST(vector<int>& nums, int lo, int hi) {
        if (hi < lo)
            return nullptr;
        int mid = (lo + hi) / 2;
        TreeNode* newTreeNode = new TreeNode(nums[mid]);
        newTreeNode->left = sortedArrayToBST(nums, lo, mid-1);
        newTreeNode->right = sortedArrayToBST(nums, mid+1, hi);
        return newTreeNode;
    }
};